### DISCUSSION TITLE
Using ADB Commands in Automations to Emulate FIremote

### DISCUSSION THREAD

**Discussion by mizuxe** [PRProd/HA-Firemote#338] [Q&A] [answered]
This may be the wrong place to ask, and I don't want to post a question that is irrelevant, so just let me know if it's a problem.  That said, I have an automation that when triggered uses ADB to pause or play my fire tv, and I wanted to add another to increase or decrease the volume just like the firemote does.  But I cannot for the life of me figure out the ADB code used.  I've been digging through the ADB github, but can't even find a list of commands and Home Assistant docs are lacking command documentation outside of pause and play. 

Is there documentation for the commands each button triggers on Firemote, or is there a database of commands and their variables? Otherwise can you just share the ADB commands that are used for volume up/volume down buttons?

**Comment by PRProd** [comment, marked as answer]
@mizuxe ~ This is the PERFECT place to ask!

I have spent a great deal of time trying to do this same exercise for every button on every remote for every device family and every device type, so I completely understand the effort it takes to figure it all out.  Mercifully, there tends to be a lot of overlap between many devices, so that saves some time.  Here are three ways that you might find to be helpful:

<br><br>
### Check the Firemote Wiki
For a few devices, I actually created a ADB command chart in the [Firemote Wiki](https://github.com/PRProd/HA-Firemote/wiki).  I would love to make a full chart for every device someday, but as you can imagine, that's a very large undertaking.  However, if you are lucky and you own one of these three devices, I've documented it all:
 - [Smart TV Fire TV 4 Series (2021)](https://github.com/PRProd/HA-Firemote/wiki/Smart-TV---Fire-TV-4-Series-(2021)#adb-commands)
 - [Fire TV Stick 4K Max 2nd Gen (2023)](https://github.com/PRProd/HA-Firemote/wiki/Fire-TV-Stick-4K-Max-2nd-Gen-(2023)#adb-commands)
 - [Fire TV Stick 4K 2nd Gen (2023)](https://github.com/PRProd/HA-Firemote/wiki/Fire-TV-Stick-4K-2nd-Gen-(2023)#adb-commands)

Even if you don't own one of these devices, the commands you find in the charts might work for you too!  As I stated before, there tends to be a lot of overlap, which is a good thing.

<br><br>
### Use the *Debug connection* tool
There is a handy feature in Home Assistant now that allows you to see the commands that Firemote (and any other cards) are sending to the HA core!  It might feel a little clunky to use at first, but it's worth the effort.  Here's how you use it:
 1. Open Developer Tools
 2. Click the 3 dots in the upper right hand corner and select **Debug Tools**

![image](https://github.com/PRProd/HA-Firemote/assets/25486599/f6191d5c-0427-47af-9dd8-c6579210a626)

 3. Turn on the **Debug connection** switch

 ![image](https://github.com/PRProd/HA-Firemote/assets/25486599/fbe5b5e4-97f0-436f-a909-fce7762f456e)

 4. Next, open the deveoper tools in your browser.  (see [this article](https://www.computerhope.com/issues/ch002153.htm) for help) and make sure that your console window is open.  If you are using chrome, and you don't see a console window, press the **ESC** key and it should appear.
 5. Now, press any button on your Firemote, and you will see the commands appear in the console window
 
![image](https://github.com/PRProd/HA-Firemote/assets/25486599/3eeacfcc-ffc3-48f9-a28f-598300c3f3b5)

 6. When you are all done, don't forget to turn off the **Debug connection** switch using steps 1 and 2.
 
 <br><br>
 ### Hunt through the Firemote source code
If writing code isn't your thing, it can be a little unpleasant to try to make sense of what you are looking at.  It also doesn't help that the current codebase for version 3.2.7 is 13002 lines long.  Hopefully these links will at least help you know *where* to start looking if you decide to go this route:
 - [All button press commands](https://github.com/PRProd/HA-Firemote/blob/v3.2.7/dist/HA-Firemote.js#L11486-L12430)
 - [Volume up button handler code](https://github.com/PRProd/HA-Firemote/blob/v3.2.7/dist/HA-Firemote.js#L11933-L11954)
 - [Volume down button handler code](https://github.com/PRProd/HA-Firemote/blob/v3.2.7/dist/HA-Firemote.js#L11971-L11992)
 - [Mute button handler code](https://github.com/PRProd/HA-Firemote/blob/v3.2.7/dist/HA-Firemote.js#L12026-L12044)
 
<br><br>
If these methods aren't working for you, just let me know and I'm more than happy to offer specific guidance for your specific situation.  Good luck!

-Doug

**Reply by mizuxe** [reply]
Thank you this was incredibly informative!!!

**Reply by PRProd** [reply]
Follow up ~ 

Today I released support for the [onn. family of devices](https://github.com/PRProd/HA-Firemote/wiki/onn.-Device-Support-Chart).  When I was working with the devices, I took the time to write a very detailed (likely not exhaustive) [ADB command chart](https://github.com/PRProd/HA-Firemote/wiki/onn.-Google-TV-4K-Streaming-Box-(2023)#adb-commands) that could be useful for other device types as well.
